dpkg/distroless_scanner.go

```diff
@@ -0,0 +1,155 @@
+// Package dpkg implements a package indexer for dpkg packages.
+package dpkg
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/textproto"
+	"path/filepath"
+	"runtime/trace"
+
+	"github.com/quay/zlog"
+
+	"github.com/quay/claircore"
+	"github.com/quay/claircore/indexer"
+	"github.com/quay/claircore/pkg/tarfs"
+)
+
+const (
+	distrolessName    = "distroless-dpkg"
+	distrolessKind    = "package"
+	distrolessVersion = "1"
+)
+
+var (
+	_ indexer.VersionedScanner = (*Scanner)(nil)
+	_ indexer.PackageScanner   = (*Scanner)(nil)
+)
+
+// Scanner implements the scanner.PackageScanner interface.
+//
+// This looks for directories that look like dpkg databases and examines the
+// "base" and "netbase" files it finds there.
+//
+// The zero value is ready to use.
+type DistrolessScanner struct{}
+
+// Name implements scanner.VersionedScanner.
+func (ps *DistrolessScanner) Name() string { return distrolessName }
+
+// Version implements scanner.VersionedScanner.
+func (ps *DistrolessScanner) Version() string { return distrolessVersion }
+
+// Kind implements scanner.VersionedScanner.
+func (ps *DistrolessScanner) Kind() string { return distrolessKind }
+
+// Scan attempts to find a dpkg database files in the layer and read all
+// of the installed packages it can find. These files are found in the
+// dpkg/status.d directory.
+//
+// It's expected to return (nil, nil) if there's no dpkg databases in the layer.
+//
+// It does not respect any dpkg configuration files.
+func (ps *DistrolessScanner) Scan(ctx context.Context, layer *claircore.Layer) ([]*claircore.Package, error) {
+	defer trace.StartRegion(ctx, "Scanner.Scan").End()
+	trace.Log(ctx, "layer", layer.Hash.String())
+	ctx = zlog.ContextWithValues(ctx,
+		"component", "distroless-dpkg/Scanner.Scan",
+		"version", ps.Version(),
+		"layer", layer.Hash.String())
+	zlog.Debug(ctx).Msg("start")
+	defer zlog.Debug(ctx).Msg("done")
+
+	rd, err := layer.Reader()
+	if err != nil {
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+	defer rd.Close()
+	sys, err := tarfs.New(rd)
+	if err != nil {
+		return nil, fmt.Errorf("opening layer failed: %w", err)
+	}
+
+	var pkgs []*claircore.Package
+	walk := func(p string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.Name() == "status.d" {
+			zlog.Debug(ctx).Msg(p)
+		}
+		if d.Name() == "status.d" && d.IsDir() {
+			dbFiles, err := sys.ReadDir(p)
+			if err != nil {
+				return fmt.Errorf("error reading DB directory: %w", err)
+			}
+			for _, f := range dbFiles {
+				pkgCt := 0
+				fn := filepath.Join(p, f.Name())
+				ctx = zlog.ContextWithValues(ctx, "database-file", fn)
+				zlog.Debug(ctx).Msg("examining package database")
+				db, err := sys.Open(fn)
+				switch {
+				case errors.Is(err, nil):
+				case errors.Is(err, fs.ErrNotExist):
```

Comment:
This looks like a porting artifact and should never happen because we're using the results of the `ReadDir`.